### PR TITLE
added new method get_nth_day_of_week

### PR DESCRIPTION
**Pull request by mahdix**

### CODE REVIEW

**Comment by mahdix** [comment]
Your suggestion makes sense, Matt.

File: t/object_methods.t [hunk 1]
```diff
@@ -85,3 +85,6 @@ is($datetime3->minus_time_interval('1d')->is_same_as($datetime2),  1,          '
 is($datetime1->minus_time_interval(0),                             $datetime1, 'minus_time_interval(0) yields the same object');
 is($datetime2->minus_time_interval('-1d')->is_same_as($datetime3), 1,          'minus_time_interval("-1d") yields one day ahead.');
 throws_ok { $datetime3->minus_time_interval("one") } qr/Bad format/, 'minus_time_interval("one") is not a mind-reader..';
+
+is($datetime1->get_nth_day_of_week(3, 3)->day_of_month, 21, 'nth_day_of_week(3, 3) is correct');
```
Comment (mwmiller) [inline t/object_methods.t]: Also, as long as I am here, you will want the two arguments here to be distinct.  With `(3,3)` it would be possible for someone to switch the argument order in the code and still have passing tests even though the behavior would be wrong. 

The tests also act as a sort of runnable documentation, so we should be able to look at the test and know how the call works.

Comment (mahdix) [inline t/object_methods.t]: By using day of week names instead of numbers, this is solved, I believe.

File: lib/Date/Utility.pm [hunk 2]
```diff
@@ -931,6 +925,32 @@ sub truncate_to_day {
...
+
+sub move_to_nth_dow {
+    my ($self, $nth, $dow_abb) = @_;
+    use Time::Local qw/timegm/;
+
+    $dow_abb //= 'undef';    # For nicer error reporting below.
+
+    my $dow = $day_abbrev_to_num{lc $dow_abb} or croak 'Invalid day of week. We got [' . $dow_abb . ']';
+
+    $dow %= 7;               # accept both 0 and 7 as Sunday
```
Comment (mwmiller) [inline lib/Date/Utility.pm]: This line is no longer necessary; we've set the number for them.

File: t/object_methods.t [hunk 3]
```diff
@@ -85,3 +85,11 @@ is($datetime3->minus_time_interval('1d')->is_same_as($datetime2),  1,          '
 is($datetime1->minus_time_interval(0),                             $datetime1, 'minus_time_interval(0) yields the same object');
 is($datetime2->minus_time_interval('-1d')->is_same_as($datetime3), 1,          'minus_time_interval("-1d") yields one day ahead.');
 throws_ok { $datetime3->minus_time_interval("one") } qr/Bad format/, 'minus_time_interval("one") is not a mind-reader..';
+
+subtest 'move_to_nth_dow' => sub {
+    is($datetime1->move_to_nth_dow(3, 'Wed')->day_of_month, 21, 'Third Wednesday of Dec 2012 is the 21st');
```
Comment (mwmiller) [inline t/object_methods.t]: Looking at a calendar, this is not true! The 21st is a Friday.  I will look into what is happening here.

Comment (mwmiller) [inline t/object_methods.t]: Description is wrong because I am an idiot.  `datetime1` is 2011, not 2012!

Comment (mahdix) [inline t/object_methods.t]: I changed the comment :-)

File: lib/Date/Utility.pm (hunk 2 above)
Comment (mahdix) [inline lib/Date/Utility.pm]: Removed.

**Comment by mwmiller** [comment]
:+1:

File: lib/Date/Utility.pm [hunk 4]
```diff
@@ -592,20 +590,17 @@ Returns the name of the current day. Example: Sunday
 
 =cut
 
+# 0..6: Sunday first.
+my @day_names = qw(Sunday Monday Tuesday Wednesday Thursday Friday Saturday);
+my %day_abbrev_to_num;
+foreach my $day_num (0 .. $#day_names) {
+    $day_abbrev_to_num{lc substr($day_names[$day_num], 0, 3)} = $day_num;
+}
```
Comment (torsten-deriv) [inline lib/Date/Utility.pm]: What you can do is this:

``` perl
my %day_abbrev_to_num = map {
    my $day = $day_names[$_];
    (lc($day) => $_, lc(substr $day, 0, 3) =>$_, $_ => $_);
} 0..$#day_names;
```

That way you can pass full day names, abbreviated day names and day numbers (0..6) to the function below.

File: lib/Date/Utility.pm [hunk 5]
```diff
@@ -931,6 +925,31 @@ sub truncate_to_day {
...
+
+    my $time     = timegm(0, 0, 0, 1, $self->month - 1, $self->year - 1900);
+    my $_dow     = (gmtime $time)[6];                                          # 0 - Sun .. 6 - Sat
+    my $days_add = ($dow + 7 - $_dow) % 7 + ($nth - 1) * 7;
+
+    my $result = Date::Utility->new($time + 24 * 3600 * $days_add);
+
+    return unless $result->month == $self->month;
+    return $result;
+}
```
Comment (torsten-deriv) [inline lib/Date/Utility.pm]: Perhaps, now that this function is part of this module it should also use the utilities provided by this module. Meaning no `Time::Local`, no `timegm` and no `gmtime`

File: lib/Date/Utility.pm [hunk 6]
```diff
@@ -931,6 +925,31 @@ sub truncate_to_day {
...
+
+    my $dow = $day_abbrev_to_num{lc $dow_abb} or croak 'Invalid day of week. We got [' . $dow_abb . ']';
+
+    my $time     = timegm(0, 0, 0, 1, $self->month - 1, $self->year - 1900);
+    my $_dow     = (gmtime $time)[6];                                          # 0 - Sun .. 6 - Sat
+    my $days_add = ($dow + 7 - $_dow) % 7 + ($nth - 1) * 7;
+
+    my $result = Date::Utility->new($time + 24 * 3600 * $days_add);
+
+    return unless $result->month == $self->month;
```
Comment (torsten-deriv) [inline lib/Date/Utility.pm]: Is this a safety measure against something like `$nth=19` or `$nth=-42`?

File: lib/Date/Utility.pm [hunk 7]
```diff
@@ -931,6 +925,31 @@ sub truncate_to_day {
     return $popular{$tepoch} // Date::Utility->new($tepoch);
 }
 
+=head2 move_to_nth_dow
+
+Returns undef or a Date::Utility object pointing to nth day of week of the current month.
+$dow is a three letter abbreviation of weekday names.
```
Comment (torsten-deriv) [inline lib/Date/Utility.pm]: If one only reads the documentation without looking at the code, one cannot understand how to use the method.

File: lib/Date/Utility.pm (hunk 6 above)
Comment (mahdix) [inline lib/Date/Utility.pm]: Yes.

File: lib/Date/Utility.pm (hunk 4 above)
Comment (mahdix) [inline lib/Date/Utility.pm]: Matt has implemented this. I will merge his PR into mine.

File: lib/Date/Utility.pm (hunk 7 above)
Comment (mahdix) [inline lib/Date/Utility.pm]: Done by Matt.

File: lib/Date/Utility.pm (hunk 5 above)
Comment (mahdix) [inline lib/Date/Utility.pm]: Done in Matt's PR.

File: t/object_methods.t [hunk 8]
```diff
@@ -2,86 +2,113 @@ use strict;
...
+    is($datetime1->minus_time_interval(0),                             $datetime1, 'minus_time_interval(0) yields the same object');
+    is($datetime2->minus_time_interval('-1d')->is_same_as($datetime3), 1,          'minus_time_interval("-1d") yields one day ahead.');
+    throws_ok { $datetime3->minus_time_interval("one") } qr/Bad format/, 'minus_time_interval("one") is not a mind-reader..';
+};
+
+subtest 'move_to_nth_dow' => sub {
+    is($datetime1->move_to_nth_dow(3,   'Wed')->day_of_month,      21,    'Third Wednesday of Dec 2011 is the 21st');
+    is($datetime1->move_to_nth_dow(5,   'SuNdaY'),                 undef, '... and there is no 5th Sunday that year.');
+    is($datetime1->move_to_nth_dow(-50, 'Mon'),                    undef, '... nor a -50th Monday.');
+    is($datetime1->move_to_nth_dow(105, 'Tue'),                    undef, '... nor a 105th Tuesday.');
```
Comment (torsten-deriv) [inline t/object_methods.t:106]: Good test cases would also be Feb 29 in 2000 (does not exist), in 2001 (does not exist) and in 2004 (does exist)

Comment (torsten-deriv) [inline t/object_methods.t:106]: What I tried to say is: It's perhaps also a good idea to test for cases that only just exceed the boundaries. `-50` and `105` are way off. But `0` is just out of bounds. Also, there is a 5th Wed in Sep 2015 but not in Aug 2015.

This kind of testing is not so much to make sure the current implementation works as expected but also against future refactoring.

Comment (mwmiller) [inline t/object_methods.t:106]: These particular tests were sort of as regression tests.  That is to say, I wrote failing tests against the old implementation and got them to pass with the new one.

I don't mind adding some more, but it's not clear to me where to add additional checks which will provide further assurance.  Showing a 5th Saturday with no 5th Sunday is trying to be that "edge check".

Famous last words, but it should be obvious that non-positive numbers are nonsensical, so adding more nonsense seems like overkill.

Comment (mwmiller) [inline t/object_methods.t:106]: > Good test cases would also be Feb 29 in 2000 (does not exist), in 2001 (does not exist) and in 2004 (does exist)

Indeed these would be some good checks, but I don't think they belong in this particular file.  Some failing cases in the construction make sense.

File: lib/Date/Utility.pm [hunk 9]
```diff
@@ -501,7 +499,7 @@ sub _parse_datetime_param {
         $year += ($year <= 30) ? 2000 : 1900;
     }
 
-    my $epoch = timegm_nocheck($second, $minute, $hour, $day, $month - 1, $year);
+    my $epoch = timegm($second, $minute, $hour, $day, $month - 1, $year);
```
Comment (torsten-deriv) [inline lib/Date/Utility.pm:502]: This breaks stuff like:

``` perl
Date::Utility->new('2015-02-28')->plus_time_interval('1d');
```

please add tests to make sure this works.

File: lib/Date/Utility.pm [hunk 10]
```diff
@@ -840,6 +838,35 @@ sub months_ahead {
...
+=cut
+
+sub move_to_nth_dow {
+    my ($self, $nth, $dow_abb) = @_;
+
+    $dow_abb //= 'undef';    # For nicer error reporting below.
+
+    my $dow = $days_to_num{lc $dow_abb} // croak 'Invalid day of week. We got [' . $dow_abb . ']';
+
+    my $dow_first = ($self->day_of_month - $self->day_of_week) % 7;
```
Comment (torsten-deriv) [inline lib/Date/Utility.pm]: this is wrong. The correct formula is:

``` perl
    my $dow_first = (1 + $self->day_of_week - $self->day_of_month) % 7;
```

File: lib/Date/Utility.pm (hunk 9 above)
Comment (torsten-deriv) [inline lib/Date/Utility.pm:502]: But simply changing it back to `timegm_nocheck` breaks one of the new tests. So, there is a bit more work to do.

File: lib/Date/Utility.pm [hunk 11]
```diff
@@ -840,6 +838,35 @@ sub months_ahead {
...
+    my $dom = ($dow + 7 - $dow_first) % 7 + ($nth - 1) * 7 + 1;
+
+    return try { Date::Utility->new(join '-', $self->year, $self->month, $dom) };
+}
```
Comment (torsten-deriv) [inline lib/Date/Utility.pm:868]: Given the error above, I think this function needs better testing. The following piece is just an example. Ideally the `19` is also varied for all days of a month. Also 1000 days is just an arbitrary number.

``` perl
    my $d = Date::Utility->new(Date::Utility::today->year . '-01-01 12:00');
    my @dow;
    my $M = '';
    for (0..1000) {
        unless ($M eq $d->year.$d->month) {
            $M = $d->year.$d->month;
            @dow = ();
        }
        $dow[$d->day_of_week]++;
        is +Date::Utility->new(join '-', $d->year, $d->month, 19)
                         ->move_to_nth_dow($dow[$d->day_of_week], $d->day_of_week)
                         ->date,
           $d->date,
           "$dow[$d->day_of_week]. ".$d->full_day_name." of ".$d->month_as_string.' '.$d->year.' is '.$d->date;

        $d = $d->plus_time_interval('1d');
    }
```

File: lib/Date/Utility.pm (hunk 9 above)
Comment (mwmiller) [inline lib/Date/Utility.pm:502]: > This breaks stuff like:
> Date::Utility->new('2015-02-28')->plus_time_interval('1d');

It absolutely does not.  Why would think that it would?

File: lib/Date/Utility.pm (hunk 10 above)
Comment (mwmiller) [inline lib/Date/Utility.pm]: We add the one in the next line.  But the variable as a poor name for what it does.

File: lib/Date/Utility.pm (hunk 11 above)
Comment (mwmiller) [inline lib/Date/Utility.pm:868]: I did extensive testing (not committed in favor of fast running tests) and it was producing correct results.  I am not sure what you're demonstrating here.

Comment (mwmiller) [inline lib/Date/Utility.pm:868]: Ok, there appears to be a problem in months which start on Sunday (as day 0)... I'm looking into this.

Comment (mwmiller) [inline lib/Date/Utility.pm:868]: Ok, I ran your script with `2e3-1` replacing `1000` and a `for` loop of `1..15` replacing `19`:

```
ok 29986 - 4. Monday of Jun 2020 is 2020-06-22
ok 29987 - 4. Monday of Jun 2020 is 2020-06-22
ok 29988 - 4. Monday of Jun 2020 is 2020-06-22
ok 29989 - 4. Monday of Jun 2020 is 2020-06-22
ok 29990 - 4. Monday of Jun 2020 is 2020-06-22
ok 29991 - 4. Monday of Jun 2020 is 2020-06-22
ok 29992 - 4. Monday of Jun 2020 is 2020-06-22
ok 29993 - 4. Monday of Jun 2020 is 2020-06-22
ok 29994 - 4. Monday of Jun 2020 is 2020-06-22
ok 29995 - 4. Monday of Jun 2020 is 2020-06-22
ok 29996 - 4. Monday of Jun 2020 is 2020-06-22
ok 29997 - 4. Monday of Jun 2020 is 2020-06-22
ok 29998 - 4. Monday of Jun 2020 is 2020-06-22
ok 29999 - 4. Monday of Jun 2020 is 2020-06-22
ok 30000 - 4. Monday of Jun 2020 is 2020-06-22
1..30000
```

So, this seems to be working now.  The algorithm, as written, seems too complex because it matches the way I thought about solving it.  I am going to enshrine a smaller version of the test in the suite and work on a simplification.  Thanks for the note.. my extensive testing was along the wrong axis to find the issue!

File: lib/Date/Utility.pm (hunk 9 above)
Comment (mahdix) [inline lib/Date/Utility.pm:502]: It works correctly according to my tests. I also added a unit test for this.

File: lib/Date/Utility.pm (hunk 11 above)
Comment (mahdix) [inline lib/Date/Utility.pm:868]: @mwmiller You are right. For example for Mar 2015 (Starts with Sunday), move_to_nth_dow(1, 0) will return Mar-7 while it has to be Mar-1

Comment (mahdix) [inline lib/Date/Utility.pm:868]: I will fix this.

Comment (mwmiller) [inline lib/Date/Utility.pm:868]: @mm-binary This is fixed (with tests demonstrating correctness) in the pull against your fork.

Comment (mahdix) [inline lib/Date/Utility.pm:868]: Thanks.

File: lib/Date/Utility.pm (hunk 9 above)
Comment (torsten-deriv) [inline lib/Date/Utility.pm:502]: Ok, it works for me as well. No idea what I did wrong when I tried it.
